### ISSUE TITLE
Conditional Step Definitions Folder for Component Testing

### ISSUE THREAD

**Issue by LewisCollum** [badeball/cypress-cucumber-preprocessor#700]
### Current behavior
Currently, `cypress.integrationFolder` is used to determine the path for step definitions in `lib/step-definitions.js::getStepDefinitionPatterns`, causing the error 

> Error: test/component/foo.feature is not inside /test/integration

when running component tests with `npx cypress run-ct`, since my component tests are in `cypress.componentFolder`. I have detailed a user workaround (see Configs below)

### Desired behavior
To incorporate cypress component testing, the path for step definitions could be determined as 
```js
if (configuration.cypress.testingType === "e2e") {
    // use integrationFolder
} else if (configuration.cypress.testingType === "component") {
    // use componentFolder
}
```
in `lib/step-definitions.js::getStepDefinitionPatterns`. So, when `npx cypress run-ct` runs, `componentFolder` is used, and when `npx cypress run` runs, `integrationFolder` is used.

See https://docs.cypress.io/api/cypress-api/testing-type for `testingType` details.

### Configs 
These configs produce passing component and integration tests with `npx cypress run-ct` and `npx cypress run`, respectively. Note that I added a hack in my `cypress/plugins/index.ts` to overwrite `integrationFolder` based on the `testingType` for now.

cypress.json
```json
{
  "testFiles": "**/*.feature",
  "component": {
    "testFiles": "**/*.feature"
  },
  "componentFolder": "test/component",  
  "integrationFolder": "test/integration"
}
```

package.json (snippet)
```json
"cypress-cucumber-preprocessor": {
  "nonGlobalStepDefinitions": true,
  "stepDefinitions": [
    "test/**/*.{js,ts,tsx,jsx}"
  ]
}
```

cypress/plugins/index.ts
```ts
import { startDevServer } from '@cypress/webpack-dev-server';
import webpack from 'webpack';
import webpackPreprocessor from "@cypress/webpack-preprocessor";

export default (on: Cypress.PluginEvents, config: Cypress.PluginConfigOptions): void => {
    // Workaround HERE to use component step definitions
    if (config.testingType === "component") {
	config.integrationFolder = config.componentFolder as string;
    }
    
    const webpackConfig : webpack.Configuration = {
	resolve: {
	    extensions: [".ts", ".js", ".tsx", ".jsx"],
	},
	module: {
	    rules: [
		{
		    test: /\.(ts|tsx)$/,
		    exclude: [/node_modules/],
		    use: [{ loader: "ts-loader" }],
		},
		{
		    test: /\.feature$/,
		    use: [
			{
			    loader: "@badeball/cypress-cucumber-preprocessor/webpack",
			    options: config,
			},
		    ],
		},
	    ],
	},
    };

    if (config.testingType === "component") {
	on("dev-server:start", (options) => {
	    return startDevServer({ options, webpackConfig });
	})
    } else if (config.testingType === "e2e") {
	on("file:preprocessor", webpackPreprocessor({webpackOptions: webpackConfig}));
    }
}

```

### Versions

```json
  "devDependencies": {
    "@badeball/cypress-cucumber-preprocessor": "^9.1.1",
    "@cypress/react": "^5.12.4",
    "@cypress/webpack-dev-server": "^1.8.4",
    "@cypress/webpack-preprocessor": "^5.11.1",
    "cypress": "^9.5.3",
    "cypress-react-selector": "^2.3.16",
    "ts-loader": "^9.2.8",
    "webpack": "^5.72.0"
  },
```
* **Node version**: 18.0.0

**Comment by LewisCollum** [comment]
May relate to #311

**Comment by badeball** [comment]
Thanks for the heads up! I've been trying to put together an example of component testing myself, in order to ensure continued support, but I have been having a hard time (https://github.com/cypress-io/cypress/issues/19337#issuecomment-1104421599) getting it to work even without the preprocessor. :/

**Comment by LewisCollum** [comment]
You bet! For others wanting to run component tests with this preprocessor as is, I've created an example repo that demonstrates it working for component tests with a couple minor hacks (see [cypress/plugins/index.ts](https://github.com/LewisCollum/cypress-cucumber-component-example/blob/master/cypress/plugins/index.ts)).

https://github.com/LewisCollum/cypress-cucumber-component-example
Keywords: React, Typescript, Cypress, Cucumber, Webpack
Relates to: #649

**Comment by badeball** [comment]
With Cypress v10 and removal of `integrationFolder`, component testing works pretty much out of the box and I've added an [example](https://github.com/badeball/cypress-cucumber-preprocessor/tree/master/examples/ct-react-ts) illustrating it. There's however currently no way to specify separate preprocessor configuration for each mode of testing, but maybe there will be.

**Comment by oguzhantx** [comment]
@badeball would it be possible to have an example for cypress component testing with cjs?

**Comment by nilssonj-tgc** [comment]
> You bet! For others wanting to run component tests with this preprocessor as is, I've created an example repo that demonstrates it working for component tests with a couple minor hacks (see [cypress/plugins/index.ts](https://github.com/LewisCollum/cypress-cucumber-component-example/blob/master/cypress/plugins/index.ts)).
> 
> https://github.com/LewisCollum/cypress-cucumber-component-example Keywords: React, Typescript, Cypress, Cucumber, Webpack Relates to: #649

Hey @LewisCollum @badeball Is this workaround compatible with projects not using TypeScript? Or possibly an update with v12? I’d like to place my Component tests within their respective component folders (app/component/ComponentName), if possible, rather than within cypress/component/ComponentName

That way I don’t have to recreate the folder structure and the test files are conveniently next to the component file.

Thanks for the help!

**Comment by CMadden8** [comment]
I've followed the instructions here for an Angular app rather than a React app: https://github.com/badeball/cypress-cucumber-preprocessor/tree/master/examples/ct-react-ts

The only tweak I made was to the below:

`framework: "angular"`

It works perfectly, until you try to run the feature file and it crashes with the below error:

`> Cannot find module 'C:\Users\madde\OneDrive\Desktop\github\cypress-cucumber-angular\cypress-cucumber-angular\cypress\support\component.ts'`

The stack trace reveals:

` at webpackMissingModule (http://localhost:8080/__cypress/src/cypress-entry.js:7296:84)
    at _ZoneDelegate.invoke (http://localhost:8080/__cypress/src/polyfills.js:7575:158)
    at Zone.run (http://localhost:8080/__cypress/src/polyfills.js:7358:35)
    at <unknown> (http://localhost:8080/__cypress/src/polyfills.js:8451:28)
    at _ZoneDelegate.invokeTask (http://localhost:8080/__cypress/src/polyfills.js:7602:171)
    at Zone.runTask (http://localhost:8080/__cypress/src/polyfills.js:7404:37)
    at drainMicroTaskQueue (http://localhost:8080/__cypress/src/polyfills.js:7773:23)
From previous event:
    at evalScripts (http://localhost:8080/__cypress/runner/cypress_runner.js:158861:58)
    at <unknown> (http://localhost:8080/__cypress/runner/cypress_runner.js:158870:152)
From previous event:
    at runScriptsFromUrls (http://localhost:8080/__cypress/runner/cypress_runner.js:158870:136)
    at Object.runScripts (http://localhost:8080/__cypress/runner/cypress_runner.js:158884:12)
    at $Cypress.onSpecWindow (http://localhost:8080/__cypress/runner/cypress_runner.js:147480:75)
    at init (http://localhost:8080/__cypress/src/cypress-entry.js:23:11)
    at 1462 (http://localhost:8080/__cypress/src/cypress-entry.js:7301:3)
    at __webpack_require__ (http://localhost:8080/__cypress/src/runtime.js:23:42)
    at render (http://localhost:8080/__cypress/src/cypress-entry.js:40:3)
    at 35 (http://localhost:8080/__cypress/src/cypress-entry.js:42:1)
    at __webpack_require__ (http://localhost:8080/__cypress/src/runtime.js:23:42)
    at __webpack_exec__ (http://localhost:8080/__cypress/src/cypress-entry.js:7308:48)
    at <unknown> (http://localhost:8080/__cypress/src/cypress-entry.js:7309:61)
    at webpackJsonpCallback (http://localhost:8080/__cypress/src/runtime.js:315:39)
    at <unknown> (http://localhost:8080/__cypress/src/cypress-entry.js:1:67)`

I've tried the solution from the example for a large company app, and I get the above issue.

Then I created a basic angular app with no other additions outside Cypress, @badeball/cypress-cucumber-preprocessor and @cypress/webpack-dev-server.

I get the same issue.

Does anyone know what the problem might be?

The component.ts file exists at cypress\support\component.ts in both cases. 

It has something to do with webpackConfig(), as removing this will remove the error - but obviously the test won't execute. 

I feel like I'm missing something very simple... It would be awesome to have gherkin generating the Cypress component tests!

**Comment by CMadden8** [comment]
It seems I was just missing 'ts-loader' from the above. 

However, I'm still getting the issue where it is looking for files it likely shouldn't be searching for. Eg, I can now get Angular Cypress component testing working with cucumber by manually including the files below in tsconfig.app.json:

```
  "files": [
    "src/main.ts",
    "cypress/component/app.ts",
    "cypress/support/component.ts"
  ],
```

Obviously this is incorrect, but I just wanted to see it working first. 

EDIT: It seems the issue is with how Cypress compiles the Angular app for component testing, and is for some reason adding the Cypress files to the compilation of the app. This issue only happens if the tweak is added to change the component testing to use cucumber, even though it's not clear how or why this would happen...

**Comment by CMadden8** [comment]
@badeball - just FYI that it seems this issue will also occur for React apps if they use the `files` prop in the tsconfig. I opened an issue [here](https://github.com/cypress-io/cypress/issues/26088) with the cypress team, and they acknowledged as much.

The solution to modify the ts-loader options to resolve a custom tsconfig unfortunately does not work, at least in the case of Angular. I'm going to have to park gherkin Cypress component testing, at least for now...

**Comment by badeball** [comment]
This has been fixed in v7.1.0.
